seed/seedwriter/writer.go

```diff
@@ -0,0 +1,498 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package seedwrite implements writing image seeds.
+package seedwriter
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/naming"
+)
+
+// Options holds the options for a Writer.
+type Options struct {
+	RootDir string
+	SeedDir string
+
+	DefaultChannel string
+
+	// Architecture to use if none is specified by the model,
+	// useful only for classic mode. If set must match the model otherwise.
+	Architecture string
+}
+
+// OptionSnap represents options about a snap.
+// If Name is set the snap is from the store. If Path is set the snap
+// is local at Path location.
+type OptionSnap struct {
+	Name    string
+	SnapID  string
+	Path    string
+	Channel string
+}
+
+func (s *OptionSnap) SnapName() string {
+	return s.Name
+}
+
+func (s *OptionSnap) ID() string {
+	return s.SnapID
+}
+
+var _ naming.SnapRef = (*OptionSnap)(nil)
+
+// SeedSnap holds details of a snap being added to a seed.
+type SeedSnap struct {
+	naming.SnapRef
+	Channel string
+	Path    string
+
+	Info  *snap.Info
+	ARefs []*asserts.Ref
+
+	local      bool
+	modelSnap  *asserts.ModelSnap
+	optionSnap *OptionSnap
```

Comment:
(bikesheed) I stumbled about this, I think it would be easier to read if it was named something like "optionsForSnap" or "options" or "snapOptions" to make it clear this is about options for the given snap.